bundles/binding/org.openhab.binding.expire/src/main/java/org/openhab/binding/expire/internal/ExpireGenericBindingProvider.java

```diff
@@ -73,6 +80,39 @@ public void processBindingConfiguration(String context, Item item, String bindin
         addBindingConfig(item, config);
     }
 
+    private static final List<Class<? extends State>> OPTIMAL_STATE_ORDER = new ArrayList<>();
+    static {
+        OPTIMAL_STATE_ORDER.add(UnDefType.class);
+        OPTIMAL_STATE_ORDER.add(OnOffType.class);
+        OPTIMAL_STATE_ORDER.add(DecimalType.class);
+        OPTIMAL_STATE_ORDER.add(DateTimeType.class);
+        OPTIMAL_STATE_ORDER.add(StringType.class);
+    }
+
+    /**
+     * Reorders the list of states so that most specific states are at higher position as more generic states.
+     * Example: OnOffState would be before StringState because StringState would accept "ON" or "OFF" but OnOffState is
+     * more specific.
+     *
+     * @param states
+     * @return
+     */
+    protected static List<Class<? extends State>> getStatesInOptimalOrderForParsing(
+            List<Class<? extends State>> states) {
+        List<Class<? extends State>> orderedStates = new ArrayList<>();
+
+        // Add states in optimal Order to the list
+        for (Class<? extends State> state : OPTIMAL_STATE_ORDER) {
+            if (states.remove(state)) {
+                orderedStates.add(state);
+            }
+        }
+        // Add remaining states to the list
+        orderedStates.addAll(states);
+
+        return orderedStates;
+    }
```

Comment:
It's a very elegant workaround, but IMHO it's better to work with the order of accepted data types as they are set in OH1 or ESH.  Please submit issues and pull requests to Eclipse SmartHome if there are problems or potential problems with the order (for example, the second accepted data type should be DateTimeType, but because it comes after StringType, it will never be reached).